source/extensions/transport_sockets/tls/cert_validator/spiffe/spiffe_validator.cc

```diff
@@ -269,8 +269,10 @@ size_t SPIFFEValidator::daysUntilFirstCertExpires() const {
   }
   size_t ret = SIZE_MAX;
   for (auto& cert : ca_certs_) {
-    size_t tmp = Utility::getDaysUntilExpiration(cert.get(), time_source_);
-    if (tmp < ret) {
+    int32_t tmp = Utility::getDaysUntilExpiration(cert.get(), time_source_);
+    if (tmp == -1) {
+      return tmp;
+    } else if (tmp < static_cast<int>(ret)) {
```

Comment:
A few things to consider here:
- I'm not sure what's the value of `SIZE_MAX`, but if the value is too large it could become negative, no?
- I'm also not sure if the expected return value should be max<size_t> when `getDaysUntilExpiration` is negative.
- previously `getDaysUntilExpiration` could return different negative numbers and each may change the return value. At the moment, all negative values will be considered as max<size_t>.

I think the following is similar to what you wrote, and avoids the casting unsigned->signed:
```
if (tmp < 0) {
  return SIZE_MAX;
} else if (static_cast<size_t>(tmp) < ret) {
  ret = tmp;
}
```

Reply:
I would potentially also change the API itself to return absl::optional<size_t> or similar to make it more clear if it's expired. Then call sites can actually be smarter and say something like "expired"

Reply:
Now, `getDaysUntilExpiration` returns -1 when negative which means it is expired, MAX is not necessary I think.

On the web page, we show days_until_expiration which means daysUntilFristExpiration, return -1 directly reduces compare times when multi-certs.

On my side, `-5000` and `-1` mean the same to me since all of them are expired.

Reply:
We can point it in the doc or show days_until_expiration as a string on the web page(expired when -1).

Reply:
I don't have enough context on SPIFFE, but what I'm seeing here is an early return if one of the certs has expired, and the returning a value of `static_cast<size_t>(-1)` in this case.
Also the casting of `ret` to signed int seems more error-prone than casting of non-negative `tmp` to unsigned int (after validating that it is not -1).

Also +1 to Matt's suggestion to consider modifying the return type to optional as it can avoid subtle bugs.

Reply:
Sure, I will modify the return type.